### PR TITLE
Form color change and alrt Javascript/jQuery functionality

### PR DESCRIPTION
**Pull request by lndncole**
Adding functionality to the form page with javascript/jQuery 

- form color change upon typing
- alert pop-up telling whether all fields are input or not, and if so, displaying values in fields

### CODE REVIEW

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: **Lines 3-12:**

`document.getElementById()` is vanilla javascript. Though this certainly works, you are avoiding using jQuery by doing this.

Selecting elements in the DOM is super convenient using jQuery, and can be done as follows:

```
var $nameInput = $('#name');
```

Also, it would be a bit more organized if we were to *separate our selections from our event listeners*, like this:

```
var $nameInput = $('#name');

$nameInput.on('keydown', nameInputBackgroundChange);
```

Comment (anthonypecchillo) [reply src/index.js]: Also notice how in the last code snippet above, I don't have to wrap `nameInputBackgroundChange` inside another function. I can just name the function I want to be invoked whenever the keydown event is heard by the `.on` event listener. 

jQuery 🌈

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: Keep your code `DRY` - Don't Repeat Yourself!

You've already selected the name input field elsewhere on the page. No need to do it twice! Make a variable, like I did in the comments above!

```
var $nameInput = $('#name');
```

Having this at the top of the file means that you can reference `$nameInput` anywhere else in the file.

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: ## Here is another example of non-DRY code.

### The Problem:

You've declared two functions to handle the same event (though, you are handling them as separate events...a `keydown` and a `change`).

### Listening for a single event:

It's simpler to think of this as if there is only one event happening. Let's choose to think simpler - either:

- The user is pressing a key on the keyboard.
- The value of the input field is changing.

**Therefore, we can listen for a single event.**

`keydown`, `keyup`, and `keypress` events are slightly easier to handle in jQuery, so for the sake of MVP, let's  choose to consider this event to be one of these keyboard events.

In jQuery, we can listen for and handle an event with the `.on` method. For instance, in this case you'd do:

```
$nameInput.on( _____, ______ );
```

**Read the documentation on jQuery's official docs page to determine what the two inputs should be!** 

### Handling a Single Event, with Multiple Conditions:

Though *there are not **two separate events** occurring, there are **two separate conditions** to consider:*

1. After the user presses a key, the value in the input field contains some text.
2. After the user presses a key, the value in the input field does not contain any text.

So, why not use *a single event handler function* with *a conditional statement inside*?

```  
function handleNameInputChange() {
  // A console.log here of the value in the input field will help you debug when you're listening for the wrong event
  console.log('FILL_ME_IN');

  // If the input field is empty
  if ('FILL_ME_IN') {
    // Change `background-color` to `'#4D766E'`, and `color` to `'white'`

  } else {
    // Change `background-color` to `'white'`, and `color` to `'#4D766E'` 

  }
}
```

**I'll leave the jQuery code as an exercise for you.**

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: ## Hm...also not very DRY.

### The Problem:

You have six more functions from here down, and they all do essentially the same thing...

### Challenge:

I bet you can refactor the `handleNameInputField` event handling function I pseudo-coded for you in the previous comment to be more general, so that it can handle the input field changes **for all four input fields in your form.**

**Hint:** Usually, when you want to take several highly specific functions and turn them into a single, more generalized function, the number of inputs to the function *increases*.

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: This is vanilla javascript, not jQuery. Let's stay consistent and write this entire file in jQuery!

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: It would be better for the reader if you define the functions you're using (in this case, `contactValueAlert`), *before* you invoke them.

**Review by anthonypecchillo — CHANGES_REQUESTED** [review]
## First Review: Initial Attempt at jQuery

### General Feedback:

Nice job getting things to happen on the screen! The fastest, most naive MVP solution is usually the best way to get started, and that's exactly what you've done in this PR. This is a great start.

However, there are many things that can be improved here:

- You're not using jQuery. You're using vanilla Javascript. Let's use jQuery.
- Your code is very non-DRY (DRY = Don't Repeat Yourself);
- The input field functionality is broken.
  - When the user erases all of the text from an input field, it does not revert back to its original styling.

### Next Steps:

- Read through the comments I've made in this review.
- Open up the [official docs for jQuery](https://api.jquery.com/) and use this as you code to discover the tools and syntax you need to accomplish your objectives.
- Consider re-watching the recording of our jQuery session about jQuery and Twittler.
- Refactor your code to satisfy my revision requests.

After we get the functionality working perfectly in jQuery, we'll move on to our next PR, where we will start refactoring our HTML and jQuery so that our code is more component-based.

File: src/index.js
Comment (lndncole) [reply src/index.js]: I don't understand what you mean here. 

I'm under the impression that the function is invoked when someone presses the send button. I don't understand how I could define it before it's invoked with that current understanding I have.

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: We write it this way:  jQuery

Not this way: Jquery

Comment (anthonypecchillo) [reply src/index.js]: Also, we know the variables are made with jQuery because the entire file is in jQuery.

Since what you're doing here is "selecting" elements from the DOM, let's just make this comment say:

```
// Selections
```

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: Change this comment to:

``` 
// Event Listeners:
```

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: Change this comment to:

```
// Event Handlers:
```

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: - Function names are always **camel-case**, soWeShouldNotCapitalizeTheFirstLetter.

- Also, let's be super careful how we name our variables and functions:

    1. You've written this function so beautifully, and so generally, that we can actually re-use it on ANY element that has both a `background-color` and a `color` property. (Provided they are in this green and white color scheme.)
    2. This function doesn't **just** change the background color. It also changes the text color.

Therefore, due to item (2), I think a better name for this function would be `invertInputColors`.

If you wanted to go **crazy**, you could figure out how to use jQuery to determine the **current** color for both the `background-color` and `color` properties, and then swap them in the function. If you were to do that, we would be able to use this function as described in item (1) above.

In this case, I would rename the function to `invertColors`.

**For the purposes of landing this PR, I only need you to implement `invertInputColors`, but bonus points to you if you can implement `invertColors`.**

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js:19]: Incredible use of the keyword `this` to concise-ify your code. Very impressed with this.

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: Move this to the section labelled `// Selections`

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: Move this to the section called `// Event Listeners`

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: One common pattern of naming event handler functions, like this one, is to prefix them with either `on` or `handle`.

If we choose to follow this convention, we should stay as consistent in our codebase as possible in doing so. This means you might want to update my comment above about the `invertColors` function. But for now, I think we're fine. (Worst case scenario, we write a new event handler for the `keyup` event that invokes the `invertColors` function inside.)

**For this function:** Let's rename it to `handleFormButtonClick`.

Notice the structure of the event handler naming convention:

> [on/handle][nameOfElement][eventFromEventListener]

Then, in HTML (or later, React), we can see a pattern than looks like this:

```
<button onclick="handleFormButtonClick()"></button>
```

In English:

> **On** hearing a click event on the form's button, **handle** the form button click event.

Listen...Handle...Listen...Handle...

Just like the labels of the sections in your jQuery code...

Doesn't that just tickle your soul?

Comment (anthonypecchillo) [inline src/index.js]: **Add a space between the `)` and the `{` on this line!**

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: Yikes! That's a long line of code! Let's change up our style a bit...how about this?:

```
if (
  $nameInput.val() === '' || 
  $mailInput.val() === '' || 
  $subjectInput.val() === '' || 
  $msgInput.val() === ''
) {
    // do stuff
}
```

Or this:

```
var formValues = [
  $nameInput.val(), 
  $mailInput.val(), 
  $subjectInput.val(), 
  $msgInput.val()
];

if (formValues.any(function(value) {
  return value === '';
}) {
  // do stuff
}
```

Or with ES6...

```
var formValues = [
  $nameInput.val(), 
  $mailInput.val(), 
  $subjectInput.val(), 
  $msgInput.val()
];

if (formValues.any(value => value === '') {
  // do stuff
}
```

Personally, I like the first one and the third one equally. They are equally efficient.

File: src/index.js
Comment (anthonypecchillo) [inline src/index.js]: Let's remove this comment. With a good function name, a comment to describe it is usually unnecessary.

**Review by anthonypecchillo — CHANGES_REQUESTED** [review]
## Second Review: Nowww we're jQuery'ing like the Pros

### General Feedback:

Awesome job. Really, a stellar refactor. Love it.

I provided some comments to take this from "Really Good" to "Perfect". Please make the requested revisions and re-submit so I can accept and merge. 👍

### Next Steps:

- Make the requested revisions and let's land this thing!
- Start your third PR that should contain updates previously requested for the NavBar dropdown menu.

After we get the functionality working perfectly in jQuery, we'll move on to our next PR, where we will start refactoring our HTML and jQuery so that our code is more component-based.

File: src/index.js
Comment (lndncole) [reply src/index.js]: renamed to invertInputColors - will look at the additional functionality you described later on. Doing MVP for now.

File: src/index.js
Comment (lndncole) [reply src/index.js:19]: Thank you sir!

File: src/index.js
Comment (lndncole) [reply src/index.js]: Makes sense - with that, it seems like a better name for my **invertInputColors** function might be something like _handleFormFieldBackgroundColorAndTextColorInversion_ or something like that lol. But I feel like that's too long.

File: dist/index.html
Comment (anthonypecchillo) [inline dist/index.html]: You've got a bug here! You're referring to the SVG's file path with an **absolute path** instead of a **relative path**.

This means that the path will only work on your local machine! When I run the app, I can't see the `/Users/landonmetcalf/...` directory.

Switch this to:

```
`src="./assets/other/OrnamentalDivider.svg"` 
```

and we'll be good to go!

Comment (anthonypecchillo) [reply dist/index.html]: **Uh oh!**

You also forgot to commit the `OrnamentalDivider.svg` file that you added to your repo. Can you please add the file  to the path referenced in the comment above?

File: dist/style.css
Comment (anthonypecchillo) [inline dist/style.css]: I'm noticing that when we add SVG files as images, it's not possible to update it's color with CSS. However, I've figured out how to manipulate the color in the SVG file itself. So, here's the new plan:

- Get rid of this line of code.
- Leave the color of this horizontal rule alone for now.
- When we eventually merge the code from the `landon` branch in to a `staging` or `production` branch, I'll manually manipulate the color of this asset in its SVG file.

(Actually, what I'll really do is turn it into two separate files; one for a green version, and one for a purple version, because there is another mockup with this asset being used in a purple color.)

**Review by anthonypecchillo — CHANGES_REQUESTED** [review]
## Second Review: Nitty Gritty Details - We're so close!

### General Feedback:

Great work Landon. Just a few tiny revisions to make and we are ready to merge this in.

Also, every time I `git pull`, I'm still getting merge conflicts in the code.  I think this is my fault for not having covered team-git-workflow with you guys yet. We should review our git contribution process live on a call before merging this. I'll also provide a better contributing guide doc.

### Next Steps:

- Make the requested revisions and let's land this thing!
- Start your third PR that should contain updates previously requested for the NavBar dropdown menu.

After we get the functionality working perfectly in jQuery, we'll move on to our next PR, where we will start refactoring our HTML and jQuery so that our code is more component-based.

File: src/index.js
Comment (anthonypecchillo) [reply src/index.js]: You're right on this one, we should stay consistent, but the new name you've chosen could be better. (It's way too long, and needs to follow the pattern I outlined above.)

How about: `handleFormInputKeyUp`

Comment (anthonypecchillo) [reply src/index.js]: Because `keyup` is the event that is firing.

File: src/index.js
Comment (anthonypecchillo) [reply src/index.js]: You're right, my comment was flawed here. Sorry about that!

File: dist/style.css
Comment (lndncole) [reply dist/style.css]: I figured out how to change the color as well - I forgot to remove this but it's the correct color now in the updated version. Sorry about that! Removing this line though as forgot to do so.

**Comment by anthonypecchillo** [comment]
Looks good to me! ✅

**Pull request merged**
